cpp/src/arrow/testing/json_integration.h

```diff
@@ -114,16 +100,15 @@ class ARROW_TESTING_EXPORT IntegrationJsonReader {
   /// \brief Read a particular record batch from the file
   ///
   /// \param[in] i the record batch index, does not boundscheck
-  /// \param[out] batch the read record batch
-  Status ReadRecordBatch(int i, std::shared_ptr<RecordBatch>* batch) const;
+  /// \return the record batch read
+  Result<std::shared_ptr<RecordBatch>> ReadRecordBatch(int i) const;
 
  private:
-  IntegrationJsonReader(MemoryPool* pool, const std::shared_ptr<Buffer>& data);
+  IntegrationJsonReader(MemoryPool* pool, std::shared_ptr<Buffer> data);
```

Comment:
Not for this PR, but food for future thought: a pattern which maintains access control for the constructor but enables usage of factories like `make_unique` is the private tag:

```suggestion
 private:
  class Tag {
    Tag() = default;
    friend class IntegrationJsonReader;
  };
 public:
  IntegrationJsonReader(Tag, MemoryPool* pool, std::shared_ptr<Buffer> data);
```

... probably not worth the boilerplate

Reply:
Ha. Yes, I suppose it might be nice for heavily-used public classes, but that's a lot of boilerplate for a small convenience.